fail-safely.md

```diff
@@ -0,0 +1,100 @@
+# Fail Safely
+
+Go doesn't have enums. (at least at the time of this writing and I have no idea if there are plans of adding them.)
+
+I've typically seen related values expressed as a set of constants.
+
+```go
+const (
+	Active  = "A"
+	Deleted = "D"
+	Pending = "P"
+)
+```
+
+or using iota. (no, stop it)
+
+```go
+const (
+  Active = iota + 1
+  Deleted
+  Pending
+)
+```
+
+However, in code, I may find something like:
+
+```go
+if user.State == Deleted {
+  return someErr
+}
+```
+
+which is doomed to fail open. What happens when a new state is added? Yep, it's _implicitly_ allowed. Congratulations, the new `Purged` state is now allowed where `Deleted` was not.
+
+We can make it a little better by describing what we want to continue, and negating it. See also [Happy Path Expressions](./negated-happy-path-form.md).
+
+```go
+if user.State != Active {
+  return someErr
+}
+```
+
+While this is _better_, we're still _implicitly_ denying any other type, but we're talking about a known set of values here. The best thing to do would be to exhaustively handle every related state, _explicitly_ allowing or denying each one, and by doing so, protecting against that future new state with a default deny.
+
+```go
+switch user.State {
+  case Active, Pending:
+    // okay, continue...
+
+  case Deleted:
+    return someErr
+
+  default:
+    return errNotHandled
+}
+```
+
+Go doesn't have enums, so I don't know of a way to easily check for this at compile time, but the next best thing is at runtime. At least something like this will help prevent a user getting into a weird state. Debugging that is way harder than debugging the `errNotHandled` default deny.
+
+An additional benefit is that the code will very readily point out all places that need to be considered when adding a new state, either by finding all of the locations via an editor, or at runtime when it throws errors.
+
+Do this for all related would-be enums: states, types, kinds, etc.
+
+There's one more example, and that is the explicit fail-open form, and it's not as uncommon as you might think. It's particularly useful in situations where it's still safe to proceed even if the value is unrecognized. This can be a good strategy when introducing something new and is riskier to fail closed unnecessarily than allow execution to continue.
+
+```go
+switch user.State {
+  case Active, Pending:
+    // okay, continue...
+
+  case Deleted:
```

Comment:
If you couple this approach with unit tests that check the function with all possible values of the pseudo-enum, you can catch these before production runtime.

You’d have to maintain a list of all valid values, but that’s very doable. Or use iota and loop until you hit the max value.

Reply:
Yeah, I removed a section I had about this because I felt like it distracted from the main idea. The difficult part is designing a test that will catch and fail for new values that aren't explicitly handled. You end up maintaining a duplicate set of these values for the test and if the function is one of those gnarly ones that require a ton of setup... 🤮 In some cases it starts to feel like testing a data structure. Of course, if an exhaustive test can be written, it's a good idea to do so.

The real solution is for there to be a language-level mechanic that can enforce this at compilation. Short of that, some linters claim to look for exhaustiveness; I haven't used them myself.

As the last line of defense, it's battle-tested in production, and fortunately, this pattern will throw errors that can point to exactly where the problem is.

Reply:
> The real solution is for there to be a language-level mechanic that can enforce this at compilation.

Agreed! Having exhaustiveness checking is something I take for granted I suppose, working in TypeScript primarily for the last couple years 🙃